### PR TITLE
Consumable quests party completion is now configurable (#5399)

### PR DESCRIPTION
**Pull request by rekongstor**
* Consume quests can now be completed for every party member if there's TRUE config value for partySyncConsumeQuests (optional for local hosted servers)

Tested on dedicated server + 2 clients. Won't work on "Open to LAN" server.

### CODE REVIEW

**Comment by chochem** [comment]
I dont think we need this, but I guess its just a config.

 _However_ it should only give a single reward then, right?

**Comment by rekongstor** [comment]
> _However_ it should only give a single reward then, right?

Yeah, sound fair but since all non-consumable quests have shared reward maybe it won't be that bad in this case. At least this is only an option for private servers to make consume-quests act like retrieve ones. Mostly to prevent admin-completion/multiple-consume situations. 
Also it cannot be truly fair to prevent quest rewards in case of partial submission from different party members. 
Also I've noticed _partySingleReward_ flag in quest jsons. I'm not sure if it's implemented yet but maybe it will be more appropriate solution to prevent multiple rewards for crucial quests to prevent reward duplication.

Anyway I can try to make reward only for the last player who made submission, but'll appreciate any help since I'm more in C++

**Comment by rekongstor** [comment]
Found bug with partial submission in this commit. Items are being removed from the inventory if there're less than needed+1 items in the inventory. Gonna fix this

**Comment by chochem** [comment]
well multiple rewards for a single consume would just be a cheat mode basically. You dont need to change a mod to do that.

**Comment by Dream-Master** [comment]
@rekongstor any news on this ?

**Comment by rekongstor** [comment]
> @rekongstor any news on this ?

Not yet

**Comment by rekongstor** [comment]
> @rekongstor any news on this ?

Updated dependencies for this PR since I'm playing on 2.3.2. Hope it'll merge automatically with master branch.

Tested on 2 clients + dedicated server. Sync works, partial submission works, reward is being claimed only for the last detected player.

File: src/main/java/betterquesting/questing/QuestInstance.java [hunk 1]
```diff
@@ -45,6 +47,7 @@
 public class QuestInstance implements IQuest
 {
 	private final TaskStorage tasks = new TaskStorage();
+    private boolean hasConsumableTasks = false;
```
Comment (rekongstor) [inline src/main/java/betterquesting/questing/QuestInstance.java:50]: Need to deal with spacing

File: src/main/java/betterquesting/questing/QuestInstance.java [hunk 2]
```diff
@@ -158,6 +161,15 @@ else if (!canSubmit(player))
 					{
 						done++;
 						update = true;
+                        if (hasConsumableTasks && QBConfig.partySyncConsumeQuests) {
+                            synchronized (completeUsers) {
```
Comment (rekongstor) [inline src/main/java/betterquesting/questing/QuestInstance.java:165]: It just worked. Not sure if it is okay to use _synchronized_ here

Comment (rekongstor) [reply src/main/java/betterquesting/questing/QuestInstance.java:165]: doesn't really work. consumable quests rewards are not claimable at all with this config

**Comment by Dream-Master** [comment]
@rekongstor are you still working on this or should we set it draft or close it?

**Comment by rekongstor** [comment]
> @rekongstor are you still working on this or should we set it draft or close it?

better draft it

File: src/main/java/betterquesting/questing/QuestInstance.java (hunk 2 above)
Comment (rekongstor) [reply src/main/java/betterquesting/questing/QuestInstance.java]: Looks like it works correctly on localhost but make rewards not claimable on LAN and Internet clients. I'll try to figure out but I'll appreciate any help :)

**Comment by boubou19** [comment]
@rekongstor any news on this?

**Comment by rekongstor** [comment]
> @rekongstor any news on this?

No, I've struggled to sync quests over non-local networks and didn't succeed

**Comment by boubou19** [comment]
i'm going to close this then, feel free to resubmit the PR if you succeed.

**Pull request closed without merging**